executor/src/constant_evaluator/mod.rs

```diff
@@ -1,194 +1,57 @@
-use std::{
-    collections::{BTreeMap, HashMap},
-    sync::{Arc, RwLock},
-};
-
 pub use data_structures::{get_uniquely_sized, get_uniquely_sized_cloned, VariablySizedColumn};
 use itertools::Itertools;
-use powdr_ast::{
-    analyzed::{Analyzed, Expression, FunctionValueDefinition, Symbol, TypedExpression},
-    parsed::{
-        types::{ArrayType, Type},
-        IndexAccess,
-    },
-};
-use powdr_number::{BigInt, BigUint, DegreeType, FieldElement};
-use powdr_pil_analyzer::evaluator::{self, Definitions, SymbolLookup, Value};
-use rayon::prelude::{IntoParallelIterator, ParallelIterator};
+use powdr_ast::analyzed::Analyzed;
+use powdr_number::FieldElement;
 
 mod data_structures;
+mod interpreter;
+mod jit_compiler;
 
 /// Generates the fixed column values for all fixed columns that are defined
 /// (and not just declared).
 /// @returns the names (in source order) and the values for the columns.
 /// Arrays of columns are flattened, the name of the `i`th array element
 /// is `name[i]`.
 pub fn generate<T: FieldElement>(analyzed: &Analyzed<T>) -> Vec<(String, VariablySizedColumn<T>)> {
-    let mut fixed_cols = HashMap::new();
+    let max_degree = analyzed
+        .constant_polys_in_source_order()
+        .iter()
+        .map(|(poly, _)| poly.degree.unwrap().max)
+        .max()
+        .unwrap_or_default();
+
+    let mut fixed_cols = Default::default();
+    if max_degree > (1 << 18) {
+        fixed_cols = jit_compiler::generate_values(analyzed);
+    }
     for (poly, value) in analyzed.constant_polys_in_source_order() {
         if let Some(value) = value {
             // For arrays, generate values for each index,
             // for non-arrays, set index to None.
             for (index, (name, id)) in poly.array_elements().enumerate() {
-                let index = poly.is_array().then_some(index as u64);
-                let range = poly.degree.unwrap();
-                let values = range
-                    .iter()
-                    .map(|degree| generate_values(analyzed, degree, &name, value, index))
-                    .collect::<Vec<_>>()
-                    .into();
-                assert!(fixed_cols.insert(name, (id, values)).is_none());
+                if !fixed_cols.contains_key(&(name.clone(), id)) {
```

Comment:
It would avoid one hit in the `HashMap` to use `if let hash_map::Entry::Vacant(entry) = fixed_cols.entry(...)` here.